test_util/cluster.py

```diff
@@ -278,6 +358,118 @@ def install_dcos(
         installer.postflight()
 
 
+def upgrade_dcos(cluster, installer_url, add_config_path=None):
+
+    def master_upgrade_order(cluster):
+        """Return a list of masters with the ZooKeeper leader last."""
+        return [master_zk_mode[0] for master_zk_mode in sorted(
+            cluster.master_zk_modes(),
+            # False < True
+            key=lambda host_zk_mode: host_zk_mode[1] == 'leader',
+        )]
+
+    def upgrade_host(tunnel, role, bootstrap_url):
+        # Download the install script for the new DC/OS.
+        tunnel.remote_cmd([
+            'curl',
+            '--silent',
+            '--verbose',
+            '--show-error',
+            '--fail',
+            '--location',
+            '--keepalive-time', '2',
+            '--retry', '20',
+            '--speed-limit', '100000',
+            '--speed-time', '60',
+            '--remote-name',
+            bootstrap_url + '/dcos_install.sh',
+        ])
+
+        # Remove the old DC/OS.
+        tunnel.remote_cmd(['sudo', '-i', '/opt/mesosphere/bin/pkgpanda', 'uninstall'])
+        tunnel.remote_cmd(['sudo', 'rm', '-rf', '/opt/mesosphere', '/etc/mesosphere'])
+
+        # Install the new DC/OS.
+        tunnel.remote_cmd(['sudo', 'bash', 'dcos_install.sh', '-d', role])
+
+    @retry(
+        wait_fixed=(1000 * 5),
+        stop_max_delay=(1000 * 60 * 5),
+        # Retry on SSH command error or metric not equal to expected value.
+        retry_on_exception=(lambda exc: isinstance(exc, CalledProcessError)),
+        retry_on_result=(lambda result: not result),
+    )
+    def wait_for_mesos_metric(cluster, host, key, value):
+        """Return True when host's Mesos metric key is equal to value."""
+        return cluster.mesos_metrics_snapshot(host).get(key) == value
+
+    assert all(h.public_ip for h in cluster.hosts), (
+        'All cluster hosts must be externally reachable. hosts: {}'.formation(cluster.hosts)
+    )
+    assert cluster.bootstrap_host, 'Upgrade requires a bootstrap host'
+
+    bootstrap_url = 'http://' + cluster.bootstrap_host.private_ip
+
+    logging.info('Preparing bootstrap host for upgrade')
+    with open(cluster.ssher.key_path, 'r') as key_fh:
+        ssh_key = key_fh.read()
+    installer = test_util.installer_api_test.DcosCliInstaller()
+    with cluster.ssher.tunnel(cluster.bootstrap_host) as bootstrap_host_tunnel:
+        installer.setup_remote(
+            tunnel=bootstrap_host_tunnel,
+            installer_path=cluster.ssher.home_dir + '/dcos_generate_config.sh',
+            download_url=installer_url,
+        )
+        installer.genconf(
+            bootstrap_url=bootstrap_url,
```

Comment:
nit / future thoughts: It'll be really nice when we just pass more of a config.yaml here rather than hand-blowing out some parameters.